### ISSUE TITLE
Additional comment about `ExeCmdT::ComputeAndPreload` conversion under RAW hazard

### ISSUE THREAD

**Issue by woojinnn** [kaist-cp/cs492-uarch#82]
During the today's class, there was a question about conversion into the `ExeCmdT::ComputeAndPreload` and handling RAW hazard. I don't think I answered correctly earlier, so I'd like to add an answer here.

## Question
```text
i1: Preload(b1, c1)
i2: Compute(a1, d1)
i3: Preload(b2, c2)
i4: Compute(a2, d2)
```
In this scenario, if `c1` and `b2` overlaps, can we combine `Compute(a1, d1)` and `Preload(b2, c2)` into a `ComputeAndPreload`?

## Answer
No, we can't because there will be a RAW hazard between `c1` and `b2`.
In such case, we don't combine them into a `ComputeAndPreload`. Instead, we just wrap them into `Compute`.
It would be better to modify the conversion rule as, `(Compute, Prelolad) if no_raw_hazard => ComputeAndPreload`

**Comment by JongyCysec** [comment]
@woojinnn 
The baseline code of ```cmd_decoder``` at ```execute/mod.rs``` is as below. 

```rust
fn cmd_decoder<const EX_QUEUE_LENGTH: usize>(
    cmd: Vr<GemminiCmd>,
) -> (
    Vr<(ExeCmd<EX_QUEUE_LENGTH>, ConfigS)>,
    I<
        VrH<(ExeCmd<EX_QUEUE_LENGTH>, ConfigS), (TagsInProgress, bool, HOption<ExeCmd<EX_QUEUE_LENGTH>>, U<2>)>,
        { Dep::Helpful },
    >,
)
where
    [(); clog2(EX_QUEUE_LENGTH) + 1]:,
    [(); clog2(EX_QUEUE_LENGTH + 1) + 1]:,
{
         ...
        .map(decode_cmd::<EX_QUEUE_LENGTH>)
        .filter_map_drop_with_r_inner(|cmd, er| {
            let (((_pop_count, prev_cmd), tags_in_progress, any_pending_robs), config) = er;
            let any_matmul_in_progress = tags_in_progress.any(|mesh_tag| mesh_tag.rob_id.is_some());

            let cmd_type = if cmd.cmds[0].is_some() {
                if cmd.do_config
                    && !any_matmul_in_progress
                    && !any_pending_robs
                    && !(prev_cmd.is_some_and(|prev_cmd| prev_cmd.typ == ExeCmdType::Config))
                {
                    Some(ExeCmdType::Config)
                } else if cmd.do_preloads[0] && cmd.do_computes[1] {
                    Some(ExeCmdType::Preload)
                } else if cmd.do_computes[0] && cmd.do_preloads[1] && cmd.do_computes[2] {
                    Some(ExeCmdType::ComputeAndPreload)
                } else if cmd.do_computes[0] {
                    Some(ExeCmdType::Compute)
                } else if any_matmul_in_progress && (config.dataflow == Dataflow::OS || cmd.do_config) {
                    Some(ExeCmdType::Flush)
                } else {
                    None
                }
            } else if any_matmul_in_progress && config.dataflow == Dataflow::OS {
                Some(ExeCmdType::Flush)
            } else {
                None
            };

            ...
}
```

So, in the branch generating ```ComputeAndPreload```, should its condition be modified  as below?
```rust
else if cmd.do_computes[0] && cmd.do_preloads[1] && cmd.do_computes[2] && cmd.cmds[0].cmd.inst.rd != cmd.rs1s[1]
```

Or, does just baseline code reflect such issue and doesn't need to add or modify the code?

**Comment by woojinnn** [comment]
Yes, it's in the next combinator(`filter_map_drop_with_r_inner`). [link](https://github.com/kaist-cp/hazardflow/blob/f453a2bdce3223a932985b4a8457ac0017bf85a4/hazardflow-designs/src/gemmini/execute/mod.rs#L359-L366)
@JongyCysec

**Comment by JongyCysec** [comment]
@woojinnn 
Oh, I got it. Thank you for your answer!!

**Comment by minseongg** [comment]
We have refactored the Execute Controller implementation, and you can check the integration logic here:

https://github.com/kaist-cp/hazardflow/blob/main/hazardflow-designs/src/gemmini/execute/mod.rs#L218-L279

Here, the following lines are related to this issue (L249-266):

```rust
} else if cmd_decoded.do_computes[0] && cmd_decoded.do_preloads[1] && cmd_decoded.do_computes[2] {
    let pre_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[1]);
    let mul_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[2]);
    let mul_rs2_addr = LocalAddr::from(cmd_decoded.rs2s[2]);

    let raw_hazard = tags_in_progress.any(|tag| {
        let pre_raw_hazard = tag.addr.is_same_addr(pre_rs1_addr) && !pre_rs1_addr.is_garbage();
        let mul_raw_hazard = (tag.addr.is_same_addr(mul_rs1_addr) && !mul_rs1_addr.is_garbage())
            || (tag.addr.is_same_addr(mul_rs2_addr) && !mul_rs2_addr.is_garbage());

        !tag.addr.is_garbage() && (pre_raw_hazard || mul_raw_hazard)
    });

    if !raw_hazard {
        Some(ExeCmdType::ComputeAndPreload)
    } else {
        Some(ExeCmdType::Compute)
    }
```

**Comment by minseongg** [comment]
We have refactored the implementation, you can check the updated conversion logic here (`get_exe_cmd_type` function):

https://github.com/kaist-cp/hazardflow/blob/main/hazardflow-designs/src/gemmini/execute/mod.rs#L218-L279

Here, the following lines are related to this issue (L249-266):

```rust
} else if cmd_decoded.do_computes[0] && cmd_decoded.do_preloads[1] && cmd_decoded.do_computes[2] {
    let pre_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[1]);
    let mul_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[2]);
    let mul_rs2_addr = LocalAddr::from(cmd_decoded.rs2s[2]);

    let raw_hazard = tags_in_progress.any(|tag| {
        let pre_raw_hazard = tag.addr.is_same_addr(pre_rs1_addr) && !pre_rs1_addr.is_garbage();
        let mul_raw_hazard = (tag.addr.is_same_addr(mul_rs1_addr) && !mul_rs1_addr.is_garbage())
            || (tag.addr.is_same_addr(mul_rs2_addr) && !mul_rs2_addr.is_garbage());

        !tag.addr.is_garbage() && (pre_raw_hazard || mul_raw_hazard)
    });

    if !raw_hazard {
        Some(ExeCmdType::ComputeAndPreload)
    } else {
        Some(ExeCmdType::Compute)
    }
```

**Comment by JongyCysec** [comment]
@minseongg @woojinnn 
Please forgive me for asking you one more detailed confirmation question.
```rust
} else if cmd_decoded.do_computes[0] && cmd_decoded.do_preloads[1] && cmd_decoded.do_computes[2] {
    let pre_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[1]);
    let mul_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[2]);
    let mul_rs2_addr = LocalAddr::from(cmd_decoded.rs2s[2]);

    let raw_hazard = tags_in_progress.any(|tag| {
        let pre_raw_hazard = tag.addr.is_same_addr(pre_rs1_addr) && !pre_rs1_addr.is_garbage();
        let mul_raw_hazard = (tag.addr.is_same_addr(mul_rs1_addr) && !mul_rs1_addr.is_garbage())
            || (tag.addr.is_same_addr(mul_rs2_addr) && !mul_rs2_addr.is_garbage());

        !tag.addr.is_garbage() && (pre_raw_hazard || mul_raw_hazard)
    });

    if !raw_hazard {
        Some(ExeCmdType::ComputeAndPreload)
    } else {
        Some(ExeCmdType::Compute)
    }
```
(1)
So, ```pre_rs1_addr``` corresponds to ```b2``` of ```i3: Preload(b2, c2)``` and ```tag.addr``` corresponds to ```c1``` of ```i1: Preload(b1, c1)```. 
And, ```pre_raw_hazard = tag.addr.is_same_addr(pre_rs1_addr) ... ``` checks whether such RAW hazard of this issue happens or not. 

(2)
And the reason why we check ```tag.addr(c1)``` is identical to source registers of third command ```compute``` (```mul_rs1_addr```, ```mul_rs2_addr```) is that ```ComputeAndPreload``` command will generate destination address of ```preload(i3: preload)``` as ```TagInProgress``` at the mesh module while skipping to send the destination address of ```compute(i2: compute)``` to ```cmd_decoder```. 

Therefore, we need to not only the dependency between ```ComputeAndPreload``` and previous command (i1: preload) but also we need to check the dependency between ```ComputeAndPreload``` and subsequent command (i4: compute <- do_computes[2])

In other words, we need to manually checks the dependency between first ```compute``` command and third ```compute``` command.

Are those description for checking hazards(```pre_raw_hazard``` & ```mul_raw_hazard```) (1) & (2) correct?

**Comment by minseongg** [comment]
Your explanation of (1) is perfectly right.

About (2), let me give more detailed explanation of them (why we need to check third command).

It may be confusing without example.

---

Let's see how the following instruction sequence are converted into `ExeCmd` in the Cmd Decoder.

Suppose that we assume the followings:

- Config: WS dataflow, and A needs to be transposed (Computing `A1^T x B1 + D1` and `A2^T x B2 + D2`).
- Address Conflict:
  + Write address of `i1` is different with read address of `i3` (`c1` != `b2`).
  + Write address of `i1` is same as read address of `i4` (`c1` == `a2`).
- There is no matmul operation processing in the Mesh-with-delays module before `i1`.

```
i1: Preload(b1, c1)
i2: Compute(a1, d1)
i3: Preload(b2, c2)
i4: Compute(a2, d2)
```

**1. Converting `i1` to `ExeCmdType::Preload`**

```rust
} else if cmd_decoded.do_preloads[0] && cmd_decoded.do_computes[1] {
    let raw_hazard = tags_in_progress.any(|tag| {
        let pre_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[0]);
        let mul_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[1]);
        let mul_rs2_addr = LocalAddr::from(cmd_decoded.rs2s[1]);

        let pre_raw_hazard = tag.addr.is_same_addr(pre_rs1_addr) && !pre_rs1_addr.is_garbage();
        let mul_raw_hazard = (tag.addr.is_same_addr(mul_rs1_addr) && !mul_rs1_addr.is_garbage())
            || (tag.addr.is_same_addr(mul_rs2_addr) && !mul_rs2_addr.is_garbage());

        !tag.addr.is_garbage() && (pre_raw_hazard || mul_raw_hazard)
    });

    if !raw_hazard {
        Some(ExeCmdType::Preload)
    } else if config.dataflow == Dataflow::OS {
        Some(ExeCmdType::Flush)
    } else {
        None
    }
```

First, it meets the above condition (`i1` is preload, and `i2` is compute).

Then it checks the RAW hazard between these addresses:

- From (writes): `tag.addr` for each `tag` in `TagsInProgress` from Mesh-with-delays FIFO (= empty).
- To (reads):
  + `pre_rs1_addr`: first read address of `i1` (= `b1`)
  + `mul_rs1_addr`: first read address of `i2` (= `a1`)
  + `mul_rs2_addr`: second read address of `i2` (= `d1`)

Here, you might curious why we need to check hazard between tags and `i2`.

It is due to the Gemmini's synchronization logic for Transposer (Slides 17 of [Supplementary Material](https://docs.google.com/presentation/d/1hM138PSjEW0UvikWY8lPVarcwspjEsp62LJ7bUTCOd0/edit#slide=id.g318b7c04235_1_244)). By the assumption (WS + A needs to be transposed), the first read address of `i2` (= `a1`) is also need to be read by the first generated `ExeCmd`.

So, the first generated `ExeCmd` has `ExeCmdType::Preload` which reads `b1` and `a1`.

**2. Converting `i2` to `ExeCmdType::Compute`**

```rust
} else if cmd_decoded.do_computes[0] && cmd_decoded.do_preloads[1] && cmd_decoded.do_computes[2] {
    let pre_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[1]);
    let mul_rs1_addr = LocalAddr::from(cmd_decoded.rs1s[2]);
    let mul_rs2_addr = LocalAddr::from(cmd_decoded.rs2s[2]);

    let raw_hazard = tags_in_progress.any(|tag| {
        let pre_raw_hazard = tag.addr.is_same_addr(pre_rs1_addr) && !pre_rs1_addr.is_garbage();
        let mul_raw_hazard = (tag.addr.is_same_addr(mul_rs1_addr) && !mul_rs1_addr.is_garbage())
            || (tag.addr.is_same_addr(mul_rs2_addr) && !mul_rs2_addr.is_garbage());

        !tag.addr.is_garbage() && (pre_raw_hazard || mul_raw_hazard)
    });

    if !raw_hazard {
        Some(ExeCmdType::ComputeAndPreload)
    } else {
        Some(ExeCmdType::Compute)
    }
```

Next, it meets the above condition (`i2` is compute, `i3` is preload, and `i4` is compute).

Then it checks the RAW hazard between these addresses:

- From (writes): `tag.addr` for each `tag` in `TagsInProgress` from Mesh-with-delays FIFO (= `c1`).
- To (reads):
  + `pre_rs1_addr`: first read address of `i3` (= `b2`)
  + `mul_rs1_addr`: first read address of `i4` (= `a2`)
  + `mul_rs2_addr`: second read address of `i4` (= `d2`)

Here, the reason of checking hazard between tags and `i4` is also same as the above (related to Transposer). Also, the reason of it does not check hazard between tags and `i2` is that it already resolved by Step 1 (generating `Preload` exe cmd). By the assumption (`c1` == `a2`), it considers there is a RAW hazard and it generates `ExeCmd` with `ExeCmdType::Compute` which reads only `d1`.

**3 & 4. Converting `i3` to `ExeCmdType::Preload`, `i4` to `ExeCmdType::Compute`**

The remaining conversion logic is similar as above, each instruction `i3` and `i4` will be converted to `ExeCmdType::Preload` and `ExeCmdType::Compute`, respectively.

---

Sorry for the long answer, so the main reason that it also need to check hazard between tags and third instruction (`c1` and `a2` / `d2`) is that `a2` or `d2` can be read by `ComputeAndPreloaded` instruction due to the Gemmini preloading logic to the Transposer (Slide 17 of the [Supplementary Materials](https://docs.google.com/presentation/d/1hM138PSjEW0UvikWY8lPVarcwspjEsp62LJ7bUTCOd0/edit?usp=sharing)).

**Comment by JongyCysec** [comment]
@minseongg 
Oh, I see. So, the main reason why we consider the dependency between current ```preload``` command with subsequent ```compute``` command is that preloading will be overlapped with transposing of subsequent ```compute``` if possible. 

Thank you for your specific, and easy-to-understand explanations.
